source/server/hot_restarting_base.cc

```diff
@@ -0,0 +1,213 @@
+
+void HotRestartingBase::initDomainSocketAddress(sockaddr_un* address) {
+  memset(address, 0, sizeof(*address));
+  address->sun_family = AF_UNIX;
+}
+
+sockaddr_un HotRestartingBase::createDomainSocketAddress(uint64_t id, const std::string& role) {
+  // Right now we only allow a maximum of 3 concurrent envoy processes to be running. When the third
+  // starts up it will kill the oldest parent.
+  const uint64_t MAX_CONCURRENT_PROCESSES = 3;
+  id = id % MAX_CONCURRENT_PROCESSES;
+
+  // This creates an anonymous domain socket name (where the first byte of the name of \0).
+  sockaddr_un address;
+  initDomainSocketAddress(&address);
+  StringUtil::strlcpy(&address.sun_path[1],
+                      fmt::format("envoy_domain_socket_{}_{}", role, base_id_ + id).c_str(),
+                      sizeof(address.sun_path) - 1);
+  address.sun_path[0] = 0;
+  return address;
+}
+
+void HotRestartingBase::bindDomainSocket(uint64_t id, const std::string& role) {
+  Api::OsSysCalls& os_sys_calls = Api::OsSysCallsSingleton::get();
+  // This actually creates the socket and binds it. We use the socket in datagram mode so we can
+  // easily read single messages.
+  my_domain_socket_ = socket(AF_UNIX, SOCK_DGRAM | SOCK_NONBLOCK, 0);
+  sockaddr_un address = createDomainSocketAddress(id, role);
+  Api::SysCallIntResult result =
+      os_sys_calls.bind(my_domain_socket_, reinterpret_cast<sockaddr*>(&address), sizeof(address));
+  if (result.rc_ != 0) {
+    throw EnvoyException(
+        fmt::format("unable to bind domain socket with id={} (see --base-id option)", id));
+  }
+}
+
+void HotRestartingBase::sendHotRestartMessage(sockaddr_un& address,
+                                              const HotRestartMessage& proto) {
+  const uint64_t serialized_size = proto.ByteSizeLong();
+  const uint64_t total_size = sizeof(uint64_t) + serialized_size;
+  // Fill with uint64_t 'length' followed by the serialized HotRestartMessage.
+  std::vector<uint8_t> send_buf;
+  send_buf.resize(total_size);
+  *reinterpret_cast<uint64_t*>(send_buf.data()) = htobe64(serialized_size);
+  RELEASE_ASSERT(proto.SerializeWithCachedSizesToArray(send_buf.data() + sizeof(uint64_t)),
+                 "failed to serialize a HotRestartMessage");
+
+  RELEASE_ASSERT(fcntl(my_domain_socket_, F_SETFL, 0) != -1,
+                 fmt::format("Set domain socket blocking failed, errno = {}", errno));
+
+  uint8_t* next_byte_to_send = send_buf.data();
+  uint64_t sent = 0;
+  while (sent < total_size) {
+    const uint64_t cur_chunk_size = std::min(MaxSendmsgSize, total_size - sent);
+    iovec iov[1];
+    iov[0].iov_base = next_byte_to_send;
+    iov[0].iov_len = cur_chunk_size;
+    next_byte_to_send += cur_chunk_size;
+    sent += cur_chunk_size;
+    msghdr message;
+    memset(&message, 0, sizeof(message));
+    message.msg_name = &address;
+    message.msg_namelen = sizeof(address);
+    message.msg_iov = iov;
+    message.msg_iovlen = 1;
```

Comment:
Will this code be ifdef'd out somehow on (say) Windows or other environments with different networking primitives?

Reply:
Huh, good question. The existing code already uses bare `sendmsg` etc. I think compiling without hot restart available just outright doesn't build the hot_restart_impl target, and Windows must do that.